Packages/Tracking/Hands/Runtime/Scripts/HandModelManager.cs

```diff
@@ -0,0 +1,512 @@
+        public void RegisterHandModel(HandModelBase handModel, bool attemptToRegisterPair = true)
+        {
+            int handModelPairIndex = GenerateHandModelIndex(handModel, attemptToRegisterPair);
+
+            RegisterHandModel(handModel, handModelPairIndex);
+        }
+
+        /// <summary>
+        /// Registers a hand model with the provided handModelPairIndex
+        /// </summary>
+        /// <param name="handModel">The hand model to register</param>
+        /// <param name="handModelPairIndex"></param>
+        private void RegisterHandModel(HandModelBase handModel, int handModelPairIndex)
+        {
+            HandEnableDisable handEnableDisable = handModel.GetComponent<HandEnableDisable>();
+            if (handEnableDisable == null)
+            {
+                Debug.LogWarning("Registered a hand model without HandEnableDisable");
+            }
+
+            if (handModel.Handedness == Chirality.Left)
+            {
+                handModelPairs[handModelPairIndex].Left = handModel;
+                handModelPairs[handModelPairIndex].LeftEnableDisable = handEnableDisable;
+            }
+            else
+            {
+                handModelPairs[handModelPairIndex].Right = handModel;
+                handModelPairs[handModelPairIndex].RightEnableDisable = handEnableDisable;
+            }
+
+            if (handModelPairs[handModelPairIndex].HandModelPairId == null || handModelPairs[handModelPairIndex].HandModelPairId == "")
+            {
+                handModelPairs[handModelPairIndex].HandModelPairId = GetDistinctHandModelId(handModel.gameObject.transform.parent.name);
+            }
+        }
+
+        protected string GetDistinctHandModelId(string desiredHandModelId)
+        {
+            int iteration = 0;
+            var name = desiredHandModelId;
+            List<string> existingIds = handModelPairs.Select(hmp => hmp.HandModelPairId).ToList();
+
+            while (existingIds.Contains(name))
+            {
+                name = desiredHandModelId + "_[" + (iteration++) + "]";
+            }
+
+            return name;
+        }
+
+        /// <summary>
+        /// Generates a hand model index to register with the hand model list
+        /// If attempting to find a pair & the pair is already registered, uses that pair's index
+        /// Otherwise assigns the next index in the list
+        /// </summary>
+        /// <param name="handModel">the hand model to register</param>
+        /// <param name="attemptToFindPair">if true, attempts to find a pair & register it if unregistered</param>
+        /// <returns>The correct hand model index</returns>
+        private int GenerateHandModelIndex(HandModelBase handModel, bool attemptToFindPair)
+        {
+            int handModelPairIndex = handModelPairs.Count;
+
+            if (!attemptToFindPair)
+            {
+                handModelPairs.Add(new HandModelPair());
```

Comment:
I feel something is suspect here as it would end up calling handModelPairs.Add(new HandModelPair()); twice if attempToFindPair is false?

Reply:
yup good call - forgot to return in the if statement